util/gitutil/git_protocol.go

```diff
@@ -1,46 +1,73 @@
 package gitutil
 
 import (
+	"net/url"
+	"regexp"
 	"strings"
 
+	"github.com/containerd/containerd/errdefs"
 	"github.com/moby/buildkit/util/sshutil"
 )
 
+type Protocol string
+
 const (
-	HTTPProtocol = iota + 1
-	HTTPSProtocol
-	SSHProtocol
-	GitProtocol
-	UnknownProtocol
+	HTTPProtocol    Protocol = "http"
+	HTTPSProtocol   Protocol = "https"
+	SSHProtocol     Protocol = "ssh"
+	GitProtocol     Protocol = "git"
+	UnknownProtocol Protocol = "unknown"
 )
 
-// ParseProtocol parses a git URL and returns the remote url and protocol type
-func ParseProtocol(remote string) (string, int) {
-	prefixes := map[string]int{
-		"http://":  HTTPProtocol,
-		"https://": HTTPSProtocol,
-		"git://":   GitProtocol,
-		"ssh://":   SSHProtocol,
-	}
-	protocolType := UnknownProtocol
-	for prefix, potentialType := range prefixes {
-		if strings.HasPrefix(remote, prefix) {
-			remote = strings.TrimPrefix(remote, prefix)
-			protocolType = potentialType
+var supportedProtos = map[Protocol]struct{}{
+	HTTPProtocol:  {},
+	HTTPSProtocol: {},
+	SSHProtocol:   {},
+	GitProtocol:   {},
+}
+
+var protoRegexp = regexp.MustCompile(`^[a-z0-9]+://`)
+
+// ParseProtocol parses a git URL and returns the protocol type, the remote,
+// and the path.
+//
+// ParseProtocol understands implicit ssh URLs such as "git@host:repo", and
+// returns the same response as if the URL were "ssh://git@host/repo".
+func ParseProtocol(remote string) (Protocol, string, string, error) {
+	if proto := protoRegexp.FindString(remote); proto != "" {
+		proto := Protocol(strings.TrimSuffix(proto, "://"))
+		_, ok := supportedProtos[proto]
+		if !ok {
+			return proto, remote, "", errdefs.ErrInvalidArgument
 		}
-	}
 
-	if protocolType == UnknownProtocol && sshutil.IsImplicitSSHTransport(remote) {
-		protocolType = SSHProtocol
-	}
+		u, err := url.Parse(remote)
+		if err != nil {
+			return proto, remote, "", err
+		}
+		if Protocol(u.Scheme) != proto {
```

Comment:
```suggestion
		if Protocol(strings.ToLower(u.Scheme)) != proto {
```

Reply:
Not needed, since `u.Scheme` is normalized to lower case in the standard library: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/net/url/url.go;l=522